### PR TITLE
PDP-816

### PR DESCRIPTION
**Pull request by nashtapu-tibco**
Values for DP (bootstrapping) Namespace
> global:
>   tibco:
>     serviceAccount: "service-account"
>     dataPlaneId: "abcd"
> dataplaneNamespaceName: ""
> namespaces:
>   dp-namespace:
>     createBinding: true

Values for App Namespace

> global:
>   tibco:
>     serviceAccount: "service-account"
>     dataPlaneId: "abcd"
> dataplaneNamespaceName: "dp-namespace"
> namespaces:
>   app-namespace:
>     createBinding: true

### CODE REVIEW

File: charts/dp-core-rbac/values.yaml [hunk 1]
```diff
@@ -1,14 +1,12 @@
...
-  create: true
-  createBinding: true
+# details of the dataplane namespace
+dataplaneNamespaceName: ""
 
-## this creates namespaces and role-bindings in those namespaces
+# details of namespaces
 namespaces: {}
-  # nsa:
+  # dp-namespace:
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/values.yaml:11]: use the consistent naming convention and change it to 'dataplaneNamespaceName'

File: charts/dp-core-rbac/templates/_helpers.tpl [hunk 2]
```diff
@@ -49,13 +49,13 @@ Create chart name and version as used by the chart label.
 {{- define "dp-core-rbac.workloadType" }}infra{{ end -}}
 
 {{/* Cluster Role Name Prefix */}}
-{{- define "dp-core-rbac.clusterRoleNamePrefix" }}cluster-role{{ end -}}
+{{- define "dp-core-rbac.clusterRoleName" }}{{ .Values.global.tibco.dataPlaneId}}-cluster-role{{ end -}}
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/templates/_helpers.tpl:52]: resource name does not need to repeat kind of resource as it is already covered by KinD

File: charts/dp-core-rbac/templates/_helpers.tpl [hunk 3]
```diff
@@ -49,13 +49,13 @@ Create chart name and version as used by the chart label.
...
 
 {{/* Cluster Role Binding Name Prefix */}}
-{{- define "dp-core-rbac.clusterRoleBindingNamePrefix" }}cluster-role-binding{{ end -}}
+{{- define "dp-core-rbac.clusterRoleBindingName" }}{{ .Values.global.tibco.dataPlaneId}}-cluster-role-binding{{ end -}}
 
 {{/* Role Binding Name Prefix */}}
-{{- define "dp-core-rbac.roleBindingNamePrefix" }}role-binding{{ end -}}
+{{- define "dp-core-rbac.roleBindingName" }}{{ .Values.global.tibco.dataPlaneId}}-role-binding{{ end -}}
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/templates/_helpers.tpl:58]: same

File: charts/dp-core-rbac/templates/cluster-role-binding.yaml [hunk 4]
```diff
@@ -1,14 +1,14 @@
-{{- if .Values.clusterRole.createBinding }}
+{{- if eq .Values.dataplaneNamespaceName "" }}
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/templates/cluster-role-binding.yaml:1]: did you mean not empty

File: charts/dp-core-rbac/templates/cluster-role.yaml [hunk 5]
```diff
@@ -1,10 +1,8 @@
-{{- if .Values.clusterRole.create }}
----
-
+{{- if eq .Values.dataplaneNamespaceName "" }}
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/templates/cluster-role.yaml:1]: same as above

File: charts/dp-core-rbac/templates/cluster-role-binding.yaml [hunk 6]
```diff
@@ -1,14 +1,14 @@
...
   labels:
     {{- include "dp-core-rbac.labels" . | nindent 4 }}
 roleRef:
   apiGroup: rbac.authorization.k8s.io
   kind: ClusterRole
-  name: {{ include "dp-core-rbac.clusterRoleNamePrefix" . }}-1
+  name: {{ include "dp-core-rbac.clusterRoleName" . }}-cluster-scope
 subjects:
   - kind: ServiceAccount
     name: {{ .Values.global.tibco.serviceAccount }}
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/templates/cluster-role-binding.yaml:14]: I think here we should use .Values.dataplaneNamespaceName and not release namespace

File: charts/dp-core-rbac/templates/sa.yaml [hunk 7]
```diff
@@ -1,3 +1,4 @@
+{{- if eq .Values.dataplaneNamespaceName "" }}
 {{- if ne .Values.global.tibco.serviceAccount "" }}
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/templates/sa.yaml:2]: if customer has provided service account then we should not create it or it means something else?

what if customer does not provide this value?

File: charts/dp-core-rbac/templates/role-binding.yaml [hunk 8]
```diff
@@ -6,18 +6,22 @@
...
   labels:
     {{- include "dp-core-rbac.labels" $root | nindent 4 }}
 roleRef:
   apiGroup: rbac.authorization.k8s.io
   kind: ClusterRole
-  name: {{ include "dp-core-rbac.clusterRoleNamePrefix" $root }}-2
+  name: {{ include "dp-core-rbac.clusterRoleName" $root }}-ns-scope
 subjects:
 - kind: ServiceAccount
   name: {{ $root.Values.global.tibco.serviceAccount }}
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/templates/role-binding.yaml:19]: this line may have impact based comment on service account creation

File: charts/dp-core-rbac/Chart.yaml [hunk 9]
```diff
@@ -3,4 +3,4 @@ name: dp-core-rbac
 description: A Helm chart for creating service accounts and required RBACs in the namespace
```
Comment (pagarwal-tibco) [inline charts/dp-core-rbac/Chart.yaml:3]: it should also say network policies

File: charts/dp-core-rbac/values.yaml (hunk 1 above)
Comment (nashtapu-tibco) [reply charts/dp-core-rbac/values.yaml]: We are following the pattern

```
global:
  tibco:
    subscriptionId: "sub1"
    dataplaneId: "abcd"
    primaryNamespaceName: "dp-namespace"
    serviceAccount: "sa" 

createNetworkPolicy: true
additionalNetworkPolicy: {}
```

With which we can use the same values set for both the namespaces.

File: charts/dp-core-rbac/templates/_helpers.tpl (hunk 2 above)
Comment (nashtapu-tibco) [reply charts/dp-core-rbac/templates/_helpers.tpl]: Removed redundant resource kind from name.

File: charts/dp-core-rbac/templates/cluster-role-binding.yaml (hunk 4 above)
Comment (nashtapu-tibco) [reply charts/dp-core-rbac/templates/cluster-role-binding.yaml]: This has been changed to required param now which ensures that if the value is not present, helm errors out:

e.g. 

> 
> Error: UPGRADE FAILED: execution error at (dp-configure-namespace/templates/role-binding.yaml:15:16): global.tibco.primaryNamespaceName is required
> helm.go:84: [debug] execution error at (dp-configure-namespace/templates/role-binding.yaml:15:16): global.tibco.primaryNamespaceName is required

File: charts/dp-core-rbac/templates/role-binding.yaml (hunk 8 above)
Comment (nashtapu-tibco) [reply charts/dp-core-rbac/templates/role-binding.yaml]: We can use the flag createServiceAccount to use existing serviceAccount or creating a new one.

File: charts/dp-core-rbac/Chart.yaml (hunk 9 above)
Comment (nashtapu-tibco) [reply charts/dp-core-rbac/Chart.yaml]: chart name changed to dp-configure-namespace.

File: charts/dp-core-rbac/templates/_helpers.tpl [hunk 10]
```diff
@@ -49,13 +49,13 @@ Create chart name and version as used by the chart label.
...
 
 
 {{/*
```
[the remark this answers is not in the document]
Comment (nashtapu-tibco) [reply charts/dp-core-rbac/templates/_helpers.tpl:52]: Label for workload-type removed.

File: charts/dp-core-rbac/templates/sa.yaml (hunk 7 above)
Comment (nashtapu-tibco) [reply charts/dp-core-rbac/templates/sa.yaml]: Customer can use an existing service account.

File: charts/dp-configure-namespace/README.md [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+createServiceAccount: true
+
+createNetworkPolicy: true
+additionalNetworkPolicy: {}
+
+## Use-case 2:
+
+Customer has created new namespaces but the service-account, cluster-roles, cluster-role-binidng are already present and new role-binding has to be created for the new namespace.
+
+# set of global values to obtain the details of service account, dataplane id and primary namespace
```
Comment (pagarwal-tibco) [inline charts/dp-configure-namespace/README.md:22]: There is no difference visible in recipe compared to above usecase. I think we should mention the relevance of helm release namespace

File: charts/dp-configure-namespace/templates/cluster-role-binding.yaml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+{{- if eq .Release.Namespace .Values.global.tibco.primaryNamespaceName -}}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: {{ required "global.tibco.dataplaneId is required" .Values.global.tibco.dataplaneId }}
```
Comment (pagarwal-tibco) [inline charts/dp-configure-namespace/templates/cluster-role-binding.yaml:5]: can we move these validations above at one place

File: charts/dp-configure-namespace/templates/role-binding.yaml [hunk 13]
```diff
@@ -0,0 +1,15 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (pagarwal-tibco) [inline charts/dp-configure-namespace/templates/role-binding.yaml:1]: we would need a condition to only create in app namespace, right?

File: charts/dp-configure-namespace/values.yaml [hunk 14]
```diff
@@ -0,0 +1,18 @@
+global:
+  tibco:
+    subscriptionId: "sub1"
```
Comment (pagarwal-tibco) [inline charts/dp-configure-namespace/values.yaml:3]: is this used?

File: charts/dp-configure-namespace/templates/role-binding.yaml (hunk 13 above)
Comment (nashtapu-tibco) [reply charts/dp-configure-namespace/templates/role-binding.yaml:1]: As confirmed with Leena, we need to keep the role-bindings in both the primary namespace as well, as there is a probability that user could deploy the app in primary namespace.

File: charts/dp-configure-namespace/README.md (hunk 11 above)
Comment (nashtapu-tibco) [reply charts/dp-configure-namespace/README.md:33]: Covered in the README.md now.

File: charts/dp-configure-namespace/templates/cluster-role-binding.yaml (hunk 12 above)
Comment (nashtapu-tibco) [reply charts/dp-configure-namespace/templates/cluster-role-binding.yaml]: Moved under _helpers.tpl now.

File: charts/dp-configure-namespace/values.yaml (hunk 14 above)
Comment (nashtapu-tibco) [reply charts/dp-configure-namespace/values.yaml]: Removed in the new commit.

**Review by pagarwal-tibco — APPROVED** [review]

**Pull request merged**